File: backend/src/security/rateLimit.ts
// --------------------------------------------------------------------------
// A simple in-memory fixed-window rate limiter. Deliberately zero external
// dependencies, consistent with the rest of this backend.
//
// Known real limitation: this state lives in one process's memory. Behind
// a load balancer with multiple instances, each instance enforces its own
// limit independently, so the *effective* limit is (per-instance limit) x
// (instance count) — not a hard global cap. That's fine for a single-
// instance deployment (most small pilots), not fine once you scale
// horizontally. At that point, move this to a shared store (Redis) or use
// your platform's built-in rate limiting (Cloudflare, Render, etc.) instead.
// --------------------------------------------------------------------------

interface Bucket {
  count: number;
  resetAt: number;
}

const buckets = new Map<string, Bucket>();

// Periodic cleanup so this Map doesn't grow forever with one-off callers.
setInterval(() => {
  const now = Date.now();
  for (const [key, bucket] of buckets) {
    if (bucket.resetAt < now) buckets.delete(key);
  }
}, 5 * 60 * 1000).unref?.();

export interface RateLimitResult {
  allowed: boolean;
  remaining: number;
  resetAt: number;
}

export function checkRateLimit(key: string, maxRequests: number, windowMs: number): RateLimitResult {
  const now = Date.now();
  const existing = buckets.get(key);

  if (!existing || existing.resetAt < now) {
    const resetAt = now + windowMs;
    buckets.set(key, { count: 1, resetAt });
    return { allowed: true, remaining: maxRequests - 1, resetAt };
  }

  if (existing.count >= maxRequests) {
    return { allowed: false, remaining: 0, resetAt: existing.resetAt };
  }

  existing.count += 1;
  return { allowed: true, remaining: maxRequests - existing.count, resetAt: existing.resetAt };
}

File: backend/app/security/rate_limit.py
import time
import threading
from typing import NamedTuple, Dict

class Bucket(NamedTuple):
    count: int
    reset_at: float

_buckets: Dict[str, Bucket] = {}
_lock = threading.Lock()

class RateLimitResult(NamedTuple):
    allowed: bool
    remaining: int
    reset_at: float

def check_rate_limit(key: str, max_requests: int, window_ms: int) -> RateLimitResult:
    now = time.time()
    window_sec = window_ms / 1000.0
    
    with _lock:
        # Occasional cleanup if map gets large
        if len(_buckets) > 1000:
            for k in list(_buckets.keys()):
                if _buckets[k].reset_at < now:
                    del _buckets[k]
                    
        existing = _buckets.get(key)
        if not existing or existing.reset_at < now:
            reset_at = now + window_sec
            _buckets[key] = Bucket(count=1, reset_at=reset_at)
            return RateLimitResult(allowed=True, remaining=max_requests - 1, reset_at=reset_at)
            
        if existing.count >= max_requests:
            return RateLimitResult(allowed=False, remaining=0, reset_at=existing.reset_at)
            
        new_count = existing.count + 1
        _buckets[key] = Bucket(count=new_count, reset_at=existing.reset_at)
        return RateLimitResult(allowed=True, remaining=max_requests - new_count, reset_at=existing.reset_at)
